database/rrdset.c

```diff
@@ -331,7 +331,10 @@ void rrdset_free(RRDSET *st) {
     freez(st->exporting_flags);
 
     while(st->variables)  rrdsetvar_free(st->variables);
-    while(st->alarms)     rrdsetcalc_unlink(st->alarms);
+    struct rrdcalc_rrdset_alarm search;
+    search.st = st;
+    struct rrdcalc_rrdset_alarm *rra = (struct rrdcalc_rrdset_alarm *)avl_search_lock(&host->alarms_idx_health_name, (avl *)&search);
```

Comment:
You should move those lines inside `rrdsetcalc_unlink()` and not change the function prototype. You can get the `rrdset` from inside the function, in the first line where it does `RRDSET *st = rc->rrdset;`.